File: XMLCheck/Handler/DictionaryCache.cs
using System.Text.Json;

namespace XMLCheck.Handler
{
    public static class DictionaryCache
    {
        public static readonly Dictionary<string, string> MaQuocTich;
        public static readonly Dictionary<string, string> MaDanToc;
        public static readonly Dictionary<string, string> DoiTuongBHYT;
        public static readonly Dictionary<string, string> DoiTuongKCB;
        public static readonly Dictionary<string, string> KetQuaDTri;
        public static readonly Dictionary<string, string> MaLoaiRV;
        static DictionaryCache()
        {
            MaQuocTich = Load("dm_ma_quoctich.json");
            MaDanToc = Load("dm_ma_dantoc.json");
            DoiTuongBHYT = Load("dm_ma_doituong_bhyt.json");
            DoiTuongKCB = Load("dm_ma_doituong_kcb.json");
            KetQuaDTri = Load("dm_ket_qua_dtri.json");
            MaLoaiRV = Load("dm_ma_loai_rv.json");
        }

        private static Dictionary<string, string> Load(string fileName)
        {
            var path = Path.Combine(
                AppContext.BaseDirectory,
                "Resources",
                "Dictionaries",
                fileName
            );

            if (!File.Exists(path))
                return new Dictionary<string, string>();

            return JsonSerializer.Deserialize<Dictionary<string, string>>(
                File.ReadAllText(path)
            ) ?? new Dictionary<string, string>();
        }
    }
}

File: XMLValidator_v2/src/parsers/dictionary.cache.ts
import { Injectable, OnModuleInit, Logger } from '@nestjs/common';
import * as fs from 'fs';
import * as path from 'path';

@Injectable()
export class DictionaryCache implements OnModuleInit {
  private readonly logger = new Logger(DictionaryCache.name);

  public maQuocTich: Record<string, string> = {};
  public maDanToc: Record<string, string> = {};
  public doiTuongBHYT: Record<string, string> = {};
  public doiTuongKCB: Record<string, string> = {};
  public ketQuaDTri: Record<string, string> = {};
  public maLoaiRV: Record<string, string> = {};

  onModuleInit() {
    this.logger.log('Initializing DictionaryCache...');
    this.maQuocTich = this.load('dm_ma_quoctich.json');
    this.maDanToc = this.load('dm_ma_dantoc.json');
    this.doiTuongBHYT = this.load('dm_ma_doituong_bhyt.json');
    this.doiTuongKCB = this.load('dm_ma_doituong_kcb.json');
    this.ketQuaDTri = this.load('dm_ket_qua_dtri.json');
    this.maLoaiRV = this.load('dm_ma_loai_rv.json');
    this.logger.log('DictionaryCache initialized successfully.');
  }

  private load(fileName: string): Record<string, string> {
    const pathsToTry = [
      // When running compiled code in dist
      path.join(__dirname, '..', 'resources', 'dictionaries', fileName),
      // When running via ts-node from project root
      path.join(process.cwd(), 'src', 'resources', 'dictionaries', fileName),
      path.join(process.cwd(), 'dist', 'resources', 'dictionaries', fileName),
    ];

    for (const p of pathsToTry) {
      if (fs.existsSync(p)) {
        try {
          const content = fs.readFileSync(p, 'utf-8');
          const data = JSON.parse(content);
          this.logger.log(`Loaded dictionary ${fileName} from: ${p} (${Object.keys(data).length} items)`);
          return data;
        } catch (e) {
          this.logger.error(`Error parsing dictionary ${fileName} at ${p}:`, e);
        }
      }
    }

    this.logger.warn(`Dictionary file ${fileName} not found. Searched paths: ${pathsToTry.join(', ')}`);
    return {};
  }
}
export const DictionaryCacheInstance = new DictionaryCache();
DictionaryCacheInstance.onModuleInit();
